File: softpos-evolution-java-light/app/src/main/java/it/nexi/softpos_evo_java_light/activities/LastTransactionInsertValuesActivity.java
package it.nexi.softpos_evo_java_light.activities;

import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import it.nexi.softpos_evo_java_light.databinding.ActivityLastTransactionInsertValuesBinding;
import it.nexi.softpos_evo_java_light.domain.Constant;
import it.nexi.softpos_evo_java_light.domain.LoadingDialog;
import it.nexi.softpos_evo_java_light.services.PreferencesService;
import it.nexi.softpos_evo_java_light.services.SDKNexiService;
import it.nexi.softpos_evo_java_light.services.Utils;
import it.nexipos.app2app.data.model.GetLastTransactionData;

/**
 *
 * Nexi Payment
 *
 * Manages the insertion of values to be used for the execution
 * of the Nexi POS GetLastTransaction method
 *
 */
public class LastTransactionInsertValuesActivity extends AppCompatActivity {

    private ActivityLastTransactionInsertValuesBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        binding = ActivityLastTransactionInsertValuesBinding.inflate(getLayoutInflater());

        setContentView(binding.getRoot());

        // Retrieving the terminalId from the app preferences
        String terminalIdSaved = PreferencesService.getLastPaymentTerminalId(this);
        binding.editTextTerminalId.setText(terminalIdSaved);

        String callerTrxIdRandom = Utils.getRandomCallerTrxId();
        binding.editTextCallerTrxId.setText(callerTrxIdRandom);

        binding.btnStartLastTransaction.setOnClickListener(v -> {

            String terminalId = binding.editTextTerminalId.getText().toString();
            String callerTrxId = binding.editTextCallerTrxId.getText().toString();

            boolean isTipEnabled    = false;
            String tipAmount        = "";

            if (terminalId.isEmpty()) {

                Utils.showInfoAlert(this,"Error","'terminalId' cannot be empty");
                return;

            }

            if (callerTrxId.isEmpty()) {

                Utils.showInfoAlert(this,"Error","'callerTrxId' cannot be empty");
                return;

            }

            // Creating the data object for the sdk
            GetLastTransactionData getLastTransactionData = new GetLastTransactionData(Constant.CALLER_NAME,
                                                                                       terminalId,
                                                                                       callerTrxId,
                                                                                       isTipEnabled,
                                                                                       tipAmount
                                                                                      );

            // starts the sdk method.
            startGetLastTransaction(getLastTransactionData);

        }); // end binding.btnStartLastTransaction.setOnClickListener(v -> {

        // An Observer is instantiated to receive communication from the Nexi sdk,
        // after the sdk's method is launched.
        SDKNexiService.getOperationResult().observe(this, result -> {

            // SDKNexiService communicated the outcome of the
            // operation to the activity
            if (result == null) return;

            LoadingDialog.hide();

            if (result.isSuccess()) {

                // SDKNexiService returned a successful result.

                Toast.makeText(this, "OK: " + result.getDescription(), Toast.LENGTH_LONG).show();

                // The activity is closed and the SDK
                // launches the Nexi POS, passing it the
                // parameters to execute the requested operation.
                finish();

            } else {

                // SDKNexiService returned an error.
                Utils.showInfoAlert(this, "Error", result.getDescription());

            }

            // Resets the LiveData state to prevent the same message
            // from being forwarded multiple times
            // (for example, in the event of a smartphone orientation change).
            SDKNexiService.resetResult();

        }); // end SDKNexiService.getOperationResult().observe(this, result -> {

    } // end protected void onCreate(

    /**
     *
     * Starts the sdk method
     *
     */
    private void startGetLastTransaction(GetLastTransactionData data) {

        LoadingDialog.show(this);

        // requests the SDKNexiService to start the method
        SDKNexiService.executeGetLastTransaction(data);

    } // private void startGetLastTransaction(GetLastTransactionData data) {

} // end Activity

File: softpos-cloud-auth-demo-android-kotlin/app/src/main/java/it/nexi/softpos_evo_kotlin_01/activities/LastTransactionInsertValuesActivity.kt
package it.nexi.softpos_evo_kotlin_01.activities

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import it.nexi.softpos_evo_kotlin_01.domain.Constant
import it.nexi.softpos_evo_kotlin_01.commons.LoadingDialog
import it.nexi.softpos_evo_kotlin_01.services.SDKNexiService
import it.nexi.softpos_evo_kotlin_01.services.PreferencesServvice
import it.nexi.softpos_evo_kotlin_01.commons.showInfoAlert
import it.nexi.softpos_evo_kotlin_01.databinding.ActivityLastTransactionInsertValuesBinding
import it.nexipos.app2app.data.model.GetLastTransactionData
import kotlinx.coroutines.launch

/**
 *
 * Nexi Payment
 *
 * Manages the insertion of values to be used for the execution
 * of the Nexi POS GetLastTransaction method
 *
 */
class LastTransactionInsertValuesActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLastTransactionInsertValuesBinding

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        binding = ActivityLastTransactionInsertValuesBinding.inflate(layoutInflater)

        // Retrieving the terminalId from the app preferences
        val terminalId = PreferencesServvice.getTerminalId(this)

        binding.editTextTerminalId.setText(terminalId)

        setContentView(binding.root)

        binding.btnStartLastTransaction.setOnClickListener {

            val terminalId  = binding.editTextTerminalId.text.toString()
            val callerTrxId = binding.editTextCallerTrxId.text.toString()

            if (terminalId.isEmpty() || callerTrxId.isEmpty()) {

                showInfoAlert(
                    context = this@LastTransactionInsertValuesActivity,
                    titolo = "Mandatory fields",
                    messaggio = "terminalId and callerTrxId, cannot be empty"
                )

                return@setOnClickListener

            } //  if (terminalId.isNullOrEmpty()) {

            // Creating the data object for the sdk
            val getLastTransactionData = GetLastTransactionData(Constant.CALLER_NAME,
                                                                terminalId,
                                                                callerTrxId
                                                               )

            // starts the sdk method.
            startGetLastTransaction(getLastTransactionData)

        } //  binding.btnStartLastTransaction {

        // awaits the response from the SDKNexiService
        lifecycleScope.launch {

            SDKNexiService.operationResult.collect { result ->

                LoadingDialog.hide()

                result?.let {

                    if (it.isSuccess) {

                        Toast.makeText(this@LastTransactionInsertValuesActivity, "OK: ${it.description}", Toast.LENGTH_LONG).show()

                        // The activity is closed and the SDK
                        // launches the Nexi POS, passing it the
                        // parameters to execute the requested operation.
                        finish()

                    } else {

                        showInfoAlert(this@LastTransactionInsertValuesActivity, "Errore", it.description)

                    }

                    // Resets the LiveData state to prevent the same message
                    // from being forwarded multiple times
                    // (for example, in the event of a smartphone orientation change).
                    SDKNexiService.resetResult()

                } // result?.let {

            } // PaymentManager.operationResult.collect { result ->

        } //  lifecycleScope.launch {

    } // onCreate

    /**
     *
     * Starts the sdk method
     *
     */
    private fun startGetLastTransaction(data: GetLastTransactionData) {

        lifecycleScope.launch {

            LoadingDialog.show(this@LastTransactionInsertValuesActivity)

            // requests the SDKNexiService to start the method
            SDKNexiService.executeGetLastTransaction(data)

        } // lifecycleScope.launch {

    } // fun startGetLastTransaction

} // end Activity


